internal/fs/type_cache_test.go

```diff
@@ -0,0 +1,406 @@
+}
+
+func (t *typeCacheTestCommon) statAndConfirmIsDir(name string, isDir bool) {
+	fi, err = os.Stat(name)
+
+	ExpectEq(nil, err)
+	AssertNe(nil, fi)
+	ExpectEq(isDir, fi.IsDir())
+}
+
+func (t *typeCacheTestCommon) statAndExpectNotADirectoryError(name string) {
+	_, err = os.Stat(name)
+
+	ExpectNe(nil, err)
+	ExpectThat(err, oglematchers.Error(oglematchers.HasSubstr("not a directory")))
+}
+
+func (t *typeCacheTestCommon) testNoInsertionSupported() {
+	// Create a file object in GCS.
+	t.createObjectOnGCS(foo)
+
+	// Stat-call with file object. It should
+	// pass stat call, skip type-cache, and return type as file.
+	t.statAndConfirmIsDir(path.Join(mntDir, foo), false)
+
+	// Create a directory object in GCS with same name as the file object.
+	t.createObjectOnGCS(foo + "/")
+
+	// Stat-call with directory object. It should
+	// pass stat call, skip type-cache, and return type as directory.
+	t.statAndConfirmIsDir(path.Join(mntDir, foo)+"/", true)
+}
+
+// //////////////////////////////////////////////////////////////////////
+// Tests for TypeCacheTestWithMaxSize1MB
+// //////////////////////////////////////////////////////////////////////
+func (t *TypeCacheTestWithMaxSize1MB) TestNoEntryInitially() {
+	// Initially, without any existing object, type-cache
+	// should not contain any entry and os.Stat should fail.
+	_, err = os.Stat(path.Join(mntDir, foo))
+
+	ExpectNe(nil, err)
+	ExpectThat(err, oglematchers.Error(oglematchers.HasSubstr("no such file or directory")))
+}
+
+func (t *TypeCacheTestWithMaxSize1MB) TestStatBehavior_DirHiddenByFile() {
+	// Create a file object in GCS.
+	t.createObjectOnGCS(foo)
+
+	// Stat-call with the file object. It should
+	// pass stat call, hit type-cache and return type as file.
+	t.statAndConfirmIsDir(path.Join(mntDir, foo), false)
+
+	// Create a directory object in GCS with same name as the file object.
+	t.createObjectOnGCS(foo + "/")
+
+	// Stat-call with the directory object path. It should
+	// fail the stat call, as type-cache currently contains a file entry with the same name and
+	// type is returned as file which is not compatible with the passed trailing "/".
+	// So, this stat call should fail and return "not a directory error"
+	t.statAndExpectNotADirectoryError(path.Join(mntDir, foo) + "/")
+}
+
+func (t *TypeCacheTestWithMaxSize1MB) TestStatBehavior_FileHiddenByDir() {
+	// Create a directory object in GCS.
+	t.createObjectOnGCS(foo + "/")
+
+	// Stat-call with the directory object. It should
+	// pass and return type as directory.
+	t.statAndConfirmIsDir(path.Join(mntDir, foo)+"/", true)
+
+	// Create a file object in GCS with same name as the directory object.
+	t.createObjectOnGCS(foo)
+
+	// Stat-call with the file object should
+	// pass but it should still report it as a directory
+	// because of its old type-cache entry which says type is directory.
+	t.statAndConfirmIsDir(path.Join(mntDir, foo), true)
+}
+
+func (t *TypeCacheTestWithMaxSize1MB) TestSizeBasedEviction() {
```

Comment:
Now, how much time does it take?

Reply:
The same we discussed.. about 720-730 ms.